gonat.go

```diff
@@ -0,0 +1,177 @@
+package gonat
+
+import (
+	"net"
+	"time"
+
+	"github.com/getlantern/errors"
+	"github.com/getlantern/golog"
+	"github.com/oxtoacart/bpool"
+)
+
+const (
+	// DefaultBufferPoolSize is 10 MB
+	DefaultBufferPoolSize = 10000000
+
+	// DefaultBufferDepth is 250 packets
+	DefaultBufferDepth = 250
+
+	// DefaultIdleTimeout is 65 seconds
+	DefaultIdleTimeout = 65 * time.Second
+
+	// DefaultStatsInterval is 15 seconds
+	DefaultStatsInterval = 15 * time.Second
+
+	// MinConntrackTimeout sets a lower bound on how long we'll let conntrack entries persist
+	MinConntrackTimeout = 1 * time.Minute
+
+	// MaximumIPPacketSize is 65535 bytes
+	MaximumIPPacketSize = 65535
+)
+
+const (
+	minEphemeralPort  = 32768
+	maxEphemeralPort  = 61000 // consistent with most Linux kernels
+	numEphemeralPorts = maxEphemeralPort - minEphemeralPort
+)
+
+var (
+	log = golog.LoggerFor("gonat")
+)
+
+type Server interface {
+	// Serve starts processing packets and blocks until finished
+	Serve() error
+
+	// Count of accepted packets
+	AcceptedPackets() int
+
+	// Count of rejected packets
+	RejectedPackets() int
+
+	// Number of TCP connections being tracked
+	NumTCPConns() int
+
+	// Number of UDP connections being tracked
+	NumUDPConns() int
+
+	// Close stops the server and cleans up resources
+	Close() error
+}
+
+type Opts struct {
+	// IFName is the name of the interface to use for connecting upstream.
+	// If not specified, this will use the default interface for reaching the
+	// Internet.
+	IFName string
+
+	// IFAddr is the address to use for outbound packets. Overrides the IFName
+	// when specified.
+	IFAddr string
+
+	// BufferPool is a pool for buffers. If not provided, default to a 10MB pool.
+	// Each []byte in the buffer pool should be 65536 bytes.
+	BufferPool BufferPool
+
+	// BufferDepth specifies the number of outbound packets to buffer between
+	// stages in the send/receive pipeline. The default is 250.
+	BufferDepth int
+
+	// IdleTimeout specifies the amount of time before idle connections are
+	// automatically closed. The default is 65 seconds.
+	IdleTimeout time.Duration
+
+	// StatsInterval controls how frequently to display stats. Defaults to 15
+	// seconds.
+	StatsInterval time.Duration
+
+	// OnOutbound allows modifying outbound ip packets.
+	OnOutbound func(pkt *IPPacket)
+
+	// OnInbound allows modifying inbound ip packets. ft is the 5 tuple to
+	// which the current connection/UDP port mapping is keyed.
+	OnInbound func(pkt *IPPacket, downFT FiveTuple)
```

Comment:
Would you mind explaining the use cases for these OnIn/Outbound callbacks?  I'm pretty sure I get the idea, just curious if you had something specific in mind.

Reply:
You can do some funky stuff by messing with packets as they come in and out (inadvertently close other connections, etc.), which is what got me thinking about the why.  If the use cases are likely to be read-only operations, we could have pass the packet by value and avoid hairy issues.  On the other hand, a user of this package should probably realize that modifying in/outbound packets can have weird effects.

Reply:
The unit test is perhaps the best explanation. I use it to hardcode an upstream destination

Reply:
Got it.  I hadn't got to the tests yet (I ran them, but didn't look at the code).  Thanks!